alacritty_terminal/src/selection.rs

```diff
@@ -81,200 +82,92 @@ pub trait Dimensions {
-    {
-        let cols = grid.dimensions().col;
-        let lines = grid.dimensions().line.0 as isize;
-
-        // Normalize ordering of selected cells
-        let (mut start, mut end) = if region.start < region.end {
-            (region.start, region.end)
-        } else {
-            (region.end, region.start)
-        };
-
-        if alt_screen {
-            Selection::alt_screen_clamp(&mut start, &mut end, lines, cols)?;
-        }
-
-        let (mut start, mut end) = if start == end {
-            if let Some(end) = grid.bracket_search(start.into()) {
-                (start.into(), end)
+    pub fn to_span(&self, term: &Term, alt_screen: bool) -> Option<Span> {
+        let start_orig = self.region.start.point;
+        let start_side = self.region.start.side;
+        let end_orig = self.region.end.point;
+        let end_side = self.region.end.side;
+        let cols = term.dimensions().col;
+        let lines = term.dimensions().line.0 as isize;
+        let (mut start, mut end, start_side, end_side, start_type, end_type) =
+            if start_orig.line > end_orig.line || start_orig.line == end_orig.line && start_orig.col <= end_orig.col {
+                (end_orig, start_orig, end_side, start_side, self.end_type.clone(), self.start_type.clone())
             } else {
-                (grid.semantic_search_right(start.into()), grid.semantic_search_left(end.into()))
-            }
-        } else if start < end && start.line == end.line {
-            (grid.semantic_search_left(start.into()), grid.semantic_search_right(end.into()))
-        } else {
-            (grid.semantic_search_right(start.into()), grid.semantic_search_left(end.into()))
-        };
-
-        if start > end {
-            std::mem::swap(&mut start, &mut end);
-        }
-
-        Some(Span { start, end })
-    }
-
-    fn span_lines<G>(
-        grid: &G,
-        region: &Range<Point<isize>>,
-        initial_line: isize,
-        alt_screen: bool,
-    ) -> Option<Span>
-    where
-        G: Dimensions,
-    {
-        let cols = grid.dimensions().col;
-        let lines = grid.dimensions().line.0 as isize;
-
-        // First, create start and end points based on initial line and the grid
-        // dimensions.
-        let mut start = Point { col: cols - 1, line: initial_line };
-        let mut end = Point { col: Column(0), line: initial_line };
-
-        // Now, expand lines based on where cursor started and ended.
-        if region.start.line < region.end.line {
-            // Start is below end
-            start.line = min(start.line, region.start.line);
-            end.line = max(end.line, region.end.line);
-        } else {
-            // Start is above end
-            start.line = min(start.line, region.end.line);
-            end.line = max(end.line, region.start.line);
-        }
-
+                (start_orig, end_orig, start_side, end_side, self.start_type.clone(), self.end_type.clone())
+            };
```

Comment:
Why separate start side/point here? Seems like it just creates additional complexity.

And if we have selection type tied to start/end tied anyways, we might as well move that into `Anchor` (and probably rename it).

That way we go from 6 variable definitions down to 2.